aiogram/fsm/storage/memory.py

```diff
@@ -66,5 +66,14 @@ async def lock(self, bot: Bot, key: StorageKey) -> AsyncGenerator[None, None]:
         async with lock:
             yield
 
+        self._cleanup(key)
+
     async def close(self) -> None:
         self._locks.clear()
+
+    def _cleanup(self, key: Hashable) -> None:
+        if self._locks[key]._waiters is None:  # type: ignore[attr-defined]
```

Comment:
getting access to the protected attribute is a bad decision, but you can inherit from the base class and add the necessary public attribute, for example:

```python
from asyncio import Lock, Future
from typing import Optional, Deque


class _SimpleEventIsolationLock(Lock):
    _waiters: Optional[Deque[Future]] = None

    @property
    def has_waiters(self) -> bool:
        return bool(self._waiters)
```
or implement the dunder method `__bool__` and use lock directly in the conditions